### PR TITLE
[tune] Support Custom Resources

### PR DESCRIPTION
**Pull request by richardliaw**
Support arbitrary resource declarations in Tune.

Need to address https://github.com/ray-project/ray/issues/2875

### CODE REVIEW

File: python/ray/tune/trial.py [hunk 1]
```diff
@@ -38,40 +38,113 @@ def date_str():
...
-
     Attributes:
         cpu (float): Number of CPUs to allocate to the trial.
         gpu (float): Number of GPUs to allocate to the trial.
         extra_cpu (float): Extra CPUs to reserve in case the trial needs to
             launch additional Ray actors that use CPUs.
         extra_gpu (float): Extra GPUs to reserve in case the trial needs to
             launch additional Ray actors that use GPUs.
+        custom_resources (dict):
+        extra_custom_resources (dict):
```
Comment (richardliaw) [inline python/ray/tune/trial.py]: TODO

**Comment by richardliaw** [comment]
cc @hartikainen

File: python/ray/tune/ray_trial_executor.py [hunk 2]
```diff
@@ -229,16 +231,35 @@ def fetch_result(self, trial):
         return result
 
     def _commit_resources(self, resources):
+        committed = self._committed_resources
+        custom_resources = {
+            res: (committed.get(res) + resources.get_res_total(res))
+            for res in resources.custom_resources
+        }
+        for resource, val in committed.custom_resources.items():
+            custom_resources.setdefault(resource, val)
```
Comment (ericl) [inline python/ray/tune/ray_trial_executor.py]: This is kind of hard to read. Why not.
all_keys = set(resources.custom_resources.keys() + committed.custom_resources.keys())
custom_resources = { res: a.get(k, 0) + b.get(k, 0) for k in all_keys }
?

File: python/ray/tune/ray_trial_executor.py [hunk 3]
```diff
@@ -229,16 +231,35 @@ def fetch_result(self, trial):
...
+            custom_resources=custom_resources)
 
     def _return_resources(self, resources):
+        committed = self._committed_resources
+        custom_resources = {
+            res: (committed.get(res) - resources.get_res_total(res))
+            for res in resources.custom_resources
+        }
+        for resource, val in committed.custom_resources.items():
+            custom_resources.setdefault(resource, val)
```
Comment (ericl) [inline python/ray/tune/ray_trial_executor.py]: Same here

File: python/ray/tune/ray_trial_executor.py [hunk 4]
```diff
@@ -287,18 +319,34 @@ def debug_string(self):
         """Returns a human readable message for printing to the console."""
 
         if self._resources_initialized:
-            return "Resources requested: {}/{} CPUs, {}/{} GPUs".format(
+            status = "Resources requested: {}/{} CPUs, {}/{} GPUs".format(
                 self._committed_resources.cpu, self._avail_resources.cpu,
                 self._committed_resources.gpu, self._avail_resources.gpu)
+            customs = ",".join([
```
Comment (ericl) [inline python/ray/tune/ray_trial_executor.py]:
```suggestion
            customs = ", ".join([
```

**Review by ericl — APPROVED** [review]
One status nit.

**Pull request merged**